src/main/java/com/khutircraftubackend/product/ProductService.java

```diff
@@ -24,134 +27,191 @@
+		}
+		
+		if (request.available() != null) {
+			existingProduct.setAvailable(request.available());
+		}
+		
+		if (request.description() != null) {
+			existingProduct.setDescription(request.description());
+		}
+		
+		if (request.categoryId() != null && request.name() != null) {
+			existingProduct.setCategory(categoryRepository.findById(request.categoryId()).orElseThrow(() ->
+					new CategoryNotFoundException(CategoryExceptionMessages.CATEGORY_NOT_FOUND)));
+		}
+		
+		return productRepository.save(existingProduct);
+	}
+	
+	@Transactional
+	public ProductEntity updateProduct(Long productId, ProductUpdateRequest request,
+									   MultipartFile thumbnailImageFile, MultipartFile imageFile) throws IOException {
+		
+		ProductEntity existingProduct = findProductById(productId);
+		ProductMapper.INSTANCE.updateProductFromRequest(existingProduct, request);
+		
+		validateImageFiles(thumbnailImageFile, imageFile);
+		
+		if (!thumbnailImageFile.isEmpty()) {
+			String thumbnailImageUrl = fileConverterService.convert(thumbnailImageFile);
+			existingProduct.setThumbnailImageUrl(thumbnailImageUrl);
+		}
+		
+		if (!imageFile.isEmpty()) {
+			String imageUrl = fileConverterService.convert(imageFile);
+			existingProduct.setImageUrl(imageUrl);
+		}
+		
+		return productRepository.save(existingProduct);
+	}
+	
+	@Transactional
+	public void deleteProduct(Long productId) throws IOException {
+		
+		ProductEntity existingProduct = findProductById(productId);
+		
+		if (existingProduct.getThumbnailImageUrl() != null) {
+			String publicId = extractPublicId(existingProduct.getThumbnailImageUrl());
+			fileUploadService.deleteCloudinaryById(publicId);
+		}
+		
+		if (existingProduct.getImageUrl() != null) {
+			String publicId = extractPublicId(existingProduct.getImageUrl());
+			fileUploadService.deleteCloudinaryById(publicId);
+		}
+		
+		productRepository.delete(existingProduct);
+	}
+	
+	@Transactional
+	public void deleteAllProductsForCurrentSeller() throws IOException {
+		
+		SellerEntity currentSeller = sellerService.getCurrentSeller();
+		
+		List<ProductEntity> products = productRepository.findAllBySeller(currentSeller);
+		
+		for (ProductEntity product : products) {
+			
+			if (product.getThumbnailImageUrl() != null) {
+				String publicId = extractPublicId(product.getThumbnailImageUrl());
+				fileUploadService.deleteCloudinaryById(publicId);
+			}
+			
+			if (product.getImageUrl() != null) {
+				String publicId = extractPublicId(product.getImageUrl());
+				fileUploadService.deleteCloudinaryById(publicId);
+			}
+		}
+		productRepository.deleteBySeller(currentSeller);
+	}
+	
+	private String extractPublicId(String url) {
+		
+		String[] parts = url.split("/");
+		
+		return parts[parts.length - 1].split("\\.")[0];
+	}
+	
+	public List<ProductEntity> getProducts(int offset, int limit) {
+		
+		Pageable pageable = PageRequest.of(offset, limit);
+		
+		return productRepository.findAllBy(pageable);
+	}
```

Comment:
last couple things left to do here with "Product Domain":

- in both `ProductCreateRequest`
remove field and it validation rule:
  ```
  @NotNull(message = "Seller ID cannot be null")
  Long sellerId,
  ```
  we don't need, coz we can get it from `sellerService.getCurrentSeller()`

- after above change we don't need these lines in `createProduct()` method any more: 
  ```java
  SellerEntity seller = sellerService.getSellerId(request.sellerId());
  validateSeller(seller); // Seller is creating their brand new product, nothing to validate at this point.
  ```
the same is done by `ProductController` as
```java
@PreAuthorize("hasRole('SELLER') and @productService.canModifyProduct(#productId)")
```

Reply:
>   validateSeller(seller); // Seller is creating their brand new product, nothing to validate at this point.

What if an existing seller adds a new item?